### ISSUE TITLE
Replace `send` and `recv` flags with a more type-safe API

### ISSUE THREAD

**Issue by dcbishop** [erickt/rust-zmq#35]
Currently send/recv functions just take them as an int and there are no defines.

At the very least there should be static definitions. But would probably be better as a proper 'Flag' type (enum?).

**Comment by jdpage** [comment]
The [bitflags! macro](http://doc.rust-lang.org/bitflags/bitflags/index.html) should be able to solve this now.

**Comment by rotty** [comment]
@jdpage, @dcbishop Indeed, that area of the API is could do with refinement. As the `bitflags!` macro looks like the currently established way to achieve that, I hope to come up with a PR making use of that, unless anyone beats me to it (which I'd welcome ;-).

**Comment by rotty** [comment]
This will go in after 0.9, as there seems to be no way to still allow the use of "0" as a flag value using the bitflags crate. I guess for the sake of improved readability, this shouldn't be allowed, either. Thus this requires an API break.

**Comment by rotty** [comment]
After having pondered about this for a while, I don't think that `bitflags` is the appropriate solution here. We have different sets of permissible flags, e.g. `SNDMORE` doesn't make sense with `send_multipart()` or `recv()`. Ideally, we want to expose this at the type level, but I'm quite short of a concrete proposal yet, if anyone has ideas about this, I'm all ears!

**Comment by rotty** [comment]
My current thinking goes in the direction of eliminating, instead of exposing `SNDMORE` and `DONTWAIT`. An initial, rough sketch, and some thoughts can be found in the commits on [my wip/flagless branch](https://github.com/rotty/rust-zmq/tree/wip/flagless).

**Comment by rotty** [comment]
Besides `SNDMORE` and `DONTWAIT`, which I'm still undecided how to deal with, there is also `POLLIN`, `POLLOUT` and `POLLERR`, which should probably be turned into a `bitflags` type.

**Comment by rotty** [comment]
Just a status update: for the poll flags, I now have added a type alias in release/v0.8, see #151, this will be followed up by a switch to `bitflags` on master. Regarding `SNDMORE` and `DONTWAIT`, I've opened an issue to clarify their interaction in zeromq/libzmq#2269. I'll be stalling further "flagless" work waiting for that issue to (hopefully) reach a conclusion.

**Comment by BatmanAoD** [comment]
@rotty Any status update? Looks like v0.9 has still not been released.

**Comment by rotty** [comment]
The current status in `0.9.0` (and current master) is:

- The poll flags have been made type-safe, using the `PollEvents` bitflags type.
- Embarassingly `send` and `recv` still take plain integers.
